syntax = "proto3";

import "book.proto";
import "imdb.proto";
import "anime.proto";

enum Type {
    BOOK = 0;
    SHOW = 1;
    ANIME = 2;
	ALL = 3;
}

////////////////////////////////////////////
///////////////BASIC RESPONSES//////////////
////////////////////////////////////////////

message ItemInfoResponse {
	repeated itemInfo recommendations = 1;
}

message ItemInfo {
    int64 id = 1;
	string name = 2;
	Type type = 3;
}

message Item {
	optional BookData book = 1;
	optional IMDBData imdb = 2;
	optional AnimeData anime = 3;
}

////////////////////////////////////////////
////////////////////////////////////////////
////////////////////////////////////////////

////////////////////////////////////////////
////////////////BASIC REQUESTS//////////////
////////////////////////////////////////////

message LibPageRequest {
	int32 page = 1;
	int32 max_results = 2;
}

message RecommendationRequest {
    int64 user_id = 1;
    int32 max_results = 2;
	repeated Type types = 3;
}

message SearchByNameRequest {
	string name = 1;
    int32 max_results = 2;
	Type types = 3;
}

message SearchByCategoryRequest {
	string category = 1;
    int32 max_results = 2;
	Type types = 3;
}

message AddItemRequest {
	optional BookData book = 1;
	optional IMDBData imdb = 2;
	optional AnimeData anime = 3;
}

message ItemId {
	int64 id = 1;
}

////////////////////////////////////////////
////////////////////////////////////////////
////////////////////////////////////////////

service Library {
	rpc Library (LibPageRequest) returns (ItemInfoResponse);
    rpc Recommend (RecommendationRequest) returns (ItemInfoResponse);
	rpc AddItem (AddItemRequest) returns (Empty);
	rpc GetItem (ItemId) returns (Item);
	rpc RemoveItem (ItemId) returns (Empty);
	rpc UpdateSeenItem (ItemId) returns (Empty);
	rpc UpdateLikeItem (ItemId) returns (Empty);
	rpc SearchByName (SearchByNameRequest) returns (ItemInfoResponse);
	rpc SearchByCategory (SearchByCategoryRequest) returns (ItemInfoResponse);
}